### ISSUE TITLE
Softmax

### ISSUE THREAD

**Issue by kroggen** [Haiyang-W/TokenFormer#9]
In your code there is this:
```python
        if normalize_type == 'softmax': 
            # NOTE: softmax = exp_l1_norm
            # outputs = F.softmax(inputs, dim=dim) * inputs.shape[dim]
            nonlinear_outputs = torch.exp(inputs)
            norm_outputs = nonlinear_outputs / torch.norm(nonlinear_outputs, p=1, dim=dim, keepdim=True) * inputs.shape[dim]
            outputs = norm_outputs
```
It turns out that `softmax` uses a division by the sum of exponentials:
```python
softmax(x)_i = exp(x_i) / sum(exp(x_j))
```
But your code is using the sum of the absolute values.

The sum consider the sign of negative values:
```
 sum([-2, 1, 3]) = -2 + 1 + 3 = 2
```
While the L1 norm does not:
```
  L1([-2, 1, 3]) = |-2| + |1| + |3| = 2 + 1 + 3 = 6
```

The comment should be `modified softmax = exp_l1_norm`

It is also multiplying by the token dimension, on this case the sum of the attention scores is not 1.

**Comment by Haiyang-W** [comment]
Thanks for your careful checking. Acturally, after exp(), the value is positive, so softmax is equal to exp_l1_norm. 
I'm only multiplying by 'inputs.shape[dim]' here to balance the variance, so that we can achieve relatively good performance. If you remove this, the performance will be really bad.

**Comment by Haiyang-W** [comment]
Directly using Softmax without scaling by token dimension, the std will be very low, and the performance will be very poor.

**Comment by CongFu92** [comment]
hi Haiyang, I have a question about your comment, "Directly using Softmax without scaling by token dimension, the std will be very low", the std is referred to the std of which variable?  looking forward to your reply, thanks

BTW, impressed work!

**Comment by Haiyang-W** [comment]
> hi Haiyang, I have a question about your comment, "Directly using Softmax without scaling by token dimension, the std will be very low", the std is referred to the std of which variable? looking forward to your reply, thanks
> 
> BTW, impressed work!

The tensor of output after the softmax. May be you can try some specifical initialization to avoid it? But it may not be an easy thing. If this part can be solved, then you can directly use normal softmax in pattention module.

**Comment by EricLina** [comment]
I think the `* inputs.shape[dim]` term is related to the paramter initialization. 
Considering pattention as `norm(X@K)@V`, 
You use `nn.init_normal_()` as the init method of $K$ and $V$ , which will cause the $std(X@K) < std(X)$.
But `nn.init.xavier_normal_()` would let $std(X@K) = std(X)$, have you tried `nn.init.xavier_normal_()` and remove the   `* inputs.shape[dim]` term ?

**Comment by EricLina** [comment]
Oh, your default init method is [`xavier_normal_ ()`](https://github.com/Haiyang-W/TokenFormer/blob/4d56c73f407635e62f6df16b97dc897b4477129e/megatron/model/tokenformer.py#L144) already.

**Comment by EricLina** [comment]
I tried to replace all nn.linear module with patteton in transformer, but the result is not comparable but even worse.

To make model parameters same as the original one, I modified the learnble token number $k$ to half of the hidden dim $d/2$.

<img width="445" alt="Image" src="https://github.com/user-attachments/assets/993219f8-9e6e-48b0-8a56-ac1aa85d03fa" />

* The result shows that it doesn't work, the loss curve is worse than the baseline(`nn.linear`) , even I scale up the learnble token number $k$ from $d/2$ to $d$.

* What's more, I tried to remove the `norm` term above, the loss curve becomes comparable with the baseline.

@Haiyang-W  Could you look at it and give some suggestions or insights?
